### PR TITLE
PUBDEV-5789: simplified R user experience when facing java version issue

### PR DESCRIPTION
**Pull request by pkozelka**
- removed useless call to curl
- before starting h2o, check if the version is ok
- added xref comments to simplify consistent update if the version support rules evolve
- java version checking in R turned into blacklisting of versions 1.2-1.6, and happens on better place

This PR is a `rel-wright` cherry-pick of #2736 from `master`.

### CODE REVIEW

File: h2o-r/h2o-package/R/connection.R [hunk 1]
```diff
@@ -533,6 +531,12 @@ h2o.clusterStatus <- function() {
       }
     )
 
+  # NOTE for developers: keep the following blacklist in logically consistent with whitelist in java code - see water.H2O.checkUnsupportedJava, near line 1849
+  if (grepl("^java version \"1\\.[1-6]\\.", jver[1])) {
```
Comment (michalkurka) [inline h2o-r/h2o-package/R/connection.R]: What if the command produces an empty output, jver[1] will be an invalid reference - better to be safe

is there any harm in using `any(grepl("...", jver)`? just like the others? or check that the first line is really available in the output?

also I would move it after the other checks (they are more informative) and I would recommend the same thing as they do:

```
         "Please download the latest Java SE JDK 8 from the following URL:\n",
          "http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html")
```

There is no reason why your case should be different, IMHO.

I think adding a unit tests would be a good idea. You can turn lines 534-549 into a function `.check.java.version(jver)` and unit test this function.

File: h2o-r/h2o-package/R/connection.R [hunk 2]
```diff
@@ -190,8 +190,6 @@ h2o.init <- function(ip = "localhost", port = 54321, startH2O = TRUE, forceDL =
         print(rv$httpStatusCode)
         print(rv$curlErrorMessage)
 
-        #try a hail mary curl
```
Comment (michalkurka) [inline h2o-r/h2o-package/R/connection.R]: Why is this being removed? Doesn't seem related.

Comment (pkozelka) [reply h2o-r/h2o-package/R/connection.R]: I believe it *is* related, because part of the problem was a confusing message:
- in case of *any* error that prevents h2o from starting, it emphasises the false impression that the problem is with connection
- and when the problem really *is* just connection, it pollutes the console output with lines providing no additional information, because, in such a case, the connection problem is already mentioned right above curl output, look here:

```
> h2o.init()
Reading in config file: ./.h2oconfig

H2O is not running yet, starting it now...

Note:  In case of errors look at the following log files:
    /tmp/RtmpLuhky4/h2o_pk_started_from_r.out
    /tmp/RtmpLuhky4/h2o_pk_started_from_r.err

java version "1.6.0_45"
Java(TM) SE Runtime Environment (build 1.6.0_45-b06)
Java HotSpot(TM) 64-Bit Server VM (build 20.45-b01, mixed mode)

Starting H2O JVM and connecting: ............................................................ 
[1] "localhost"
[1] 54321
[1] TRUE
[1] -1
[1] "Failed to connect to localhost port 54321: Connection refused"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed to connect to localhost port 54321: Connection refused
[1] 7
Error in h2o.init() : H2O failed to start, stopping execution.
```
(this console output was created on `master` branch)


And these lines are what we loose by removing that curl:
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed to connect to localhost port 54321: Connection refused
[1] 7
```

I don't see anything useful here.

Also this result can hardly be different, because it comes in situation when the previous timeout has expired, and the connection is unlikely to be working the very nanosecond after timeout.

I can imagine that, if R uses some other network configuration (of proxies etc), it might fail, while this curl goes directly (or based on system configuration) and in such a case it might behave differently.
But, if that was the purpose (I hope not, would be a bit weird), it should be elaborated into more comprehensive information for the user, otherwise it is really just confusing.

File: h2o-r/h2o-package/R/connection.R (hunk 1 above)
Comment (pkozelka) [reply h2o-r/h2o-package/R/connection.R]: For empty output and moving after other checks - totally correct, I will fix this.

Also you are right that this check should not provide different output. But, are you sure that we want to hard-code in so many places the exact JDK url, instead of pointing the user to more detailed and h2o-specific information about JDK, in all these cases?
To maintain the same level of user comfort, we can put that recommended url to that doc page, WDIT?

**Comment by pkozelka** [comment]
@michalkurka I made the changes in checking that you suggested, including the error message text.
I still believe it would be better to point the user to our docs page rather than specific java download page, in all present java checks - but this can be done separately.

File: h2o-r/h2o-package/R/connection.R (hunk 2 above)
Comment (michalkurka) [reply h2o-r/h2o-package/R/connection.R]: In general it is a good idea to provide more information to debug the issue but in this case I think we can really remove it because:

- it hardcodes HTTP
- it relies on `curl` (useless on windows)
- it hardcodes 54321

This code is working. Let's remove it.
